gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/src/main/java/io/gravitee/plugin/endpoint/mqtt5/Mqtt5EndpointConnector.java

```diff
@@ -214,16 +195,59 @@ protected Completable publish(ExecutionContext ctx) {
         }
     }
 
-    private String getIdentifier(ExecutionContext ctx) {
-        String identifier = ctx.getAttribute(CONTEXT_ATTRIBUTE_MQTT5_IDENTIFIER);
-        if (identifier == null || identifier.isEmpty()) {
-            identifier = configuration.getIdentifier();
-            if (identifier == null) {
-                identifier = ctx.request().transactionId();
-                ctx.setAttribute(CONTEXT_ATTRIBUTE_MQTT5_IDENTIFIER, identifier);
-            }
+    private String generatePublisherIdentifier() {
+        return "gio-apim-publisher-" + UUID.randomUUID().toString().split("-")[0];
+    }
+
+    protected Mqtt5SubscribeBuilder.Publishes.Start.Complete<FlowableWithSingle<Mqtt5Publish, Mqtt5SubAck>> prepareMqtt5Subscribe(
+        final ExecutionContext ctx,
+        final Mqtt5RxClient rxClient,
+        final String topic
+    ) {
+        return rxClient.subscribePublishesWith().topicFilter(topic).qos(Mqtt5Publish.DEFAULT_QOS);
+    }
+
+    protected MqttConnectBuilder.Default prepareMqttConnect() {
+        return MqttConnect.DEFAULT.extend();
+    }
+
+    protected Mqtt5ClientBuilder prepareMqtt5Client(final ExecutionContext ctx) {
+        return Mqtt5Client
+            .builder()
+            .serverHost(configuration.getServerHost())
+            .serverPort(configuration.getServerPort())
+            .automaticReconnectWithDefaultConfig()
+            .addDisconnectedListener(
+                context -> {
+                    if (context.getReconnector().getAttempts() >= configuration().getReconnectAttempt()) {
+                        context.getReconnector().reconnect(false);
+                    }
+                }
+            );
+    }
+
+    protected Mqtt5PublishBuilder.Complete prepareMqtt5Publish(
+        final ExecutionContext ctx,
+        final String topic,
+        final String responseTopic,
+        final Message message
+    ) {
+        Mqtt5PublishBuilder.Complete builder = Mqtt5Publish
+            .builder()
+            .topic(overrideConfigFromMessage(topic, message, CONTEXT_ATTRIBUTE_MQTT5_TOPIC))
+            .responseTopic(overrideConfigFromMessage(responseTopic, message, CONTEXT_ATTRIBUTE_MQTT5_RESPONSE_TOPIC))
+            .retain(configuration.getProducer().isRetained())
+            .qos(Mqtt5Publish.DEFAULT_QOS)
+            .payload(message.content().getBytes())
+            .noMessageExpiry();
+        if (message.headers() != null) {
+            Mqtt5PublishBuilder.Complete finalBuilder = builder;
```

Comment:
I would have assigned builder.userProperties instead of just builder.
but not a big deal :p

Reply:
The "builder" api of hivemq client is a bit awkward. You need the builder to call applyUserProperties() on it and get back the new builder. Don't know if it could work without chaining the builder.